File: packages/typescript/test/seed-formats.test.ts
// Seed-format conformance: users load MatterVault API keys with
// @polkadot/keyring's `addFromUri`, as either a 0x-hex mini-secret or a BIP39
// mnemonic. Replay the Rust-emitted fixture (testvectors/seed_formats.json) to
// pin that the companion library derives the same account from both encodings —
// the exact dashboard↔SDK key-ingestion contract.

import { readFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { beforeAll, describe, expect, it } from "vitest";
import { Keyring } from "@polkadot/keyring";
import { cryptoWaitReady } from "@polkadot/util-crypto";
import { u8aToHex } from "@polkadot/util";

const vectorsDir = resolve(dirname(fileURLToPath(import.meta.url)), "../../../testvectors");
const { cases } = JSON.parse(readFileSync(resolve(vectorsDir, "seed_formats.json"), "utf8"));

describe("seed format conformance", () => {
  beforeAll(async () => {
    await cryptoWaitReady();
  });

  it.each(cases)("both encodings derive account $account_id_hex", (c: any) => {
    const keyring = new Keyring({ type: "sr25519" });
    const fromMnemonic = keyring.addFromUri(c.mnemonic);
    const fromHex = keyring.addFromUri(`0x${c.mini_secret_hex}`);
    expect(u8aToHex(fromMnemonic.publicKey)).toBe(`0x${c.account_id_hex}`);
    expect(u8aToHex(fromHex.publicKey)).toBe(`0x${c.account_id_hex}`);
  });
});

File: packages/go/mattervault/seed_formats_test.go
package mattervault

// Seed-format conformance: a MatterVault API key loads via gsrpc's
// signature.KeyringPairFromSecret as either a 0x-hex mini-secret or a BIP39
// mnemonic. Replay the Rust-emitted fixture (testvectors/seed_formats.json) to
// pin that both encodings derive the same account — the dashboard↔SDK
// key-ingestion contract.

import (
	"encoding/hex"
	"testing"

	"github.com/centrifuge/go-substrate-rpc-client/v4/signature"
)

func TestSeedFormatConformance(t *testing.T) {
	var v struct {
		Cases []struct {
			Mnemonic      string `json:"mnemonic"`
			MiniSecretHex string `json:"mini_secret_hex"`
			AccountIDHex  string `json:"account_id_hex"`
		} `json:"cases"`
	}
	load(t, "seed_formats.json", &v)
	for _, c := range v.Cases {
		fromMnemonic, err := signature.KeyringPairFromSecret(c.Mnemonic, 42)
		if err != nil {
			t.Fatalf("mnemonic: %v", err)
		}
		fromHex, err := signature.KeyringPairFromSecret("0x"+c.MiniSecretHex, 42)
		if err != nil {
			t.Fatalf("hex seed: %v", err)
		}
		if got := hex.EncodeToString(fromMnemonic.PublicKey); got != c.AccountIDHex {
			t.Fatalf("mnemonic pubkey = %s, want %s", got, c.AccountIDHex)
		}
		if got := hex.EncodeToString(fromHex.PublicKey); got != c.AccountIDHex {
			t.Fatalf("hex pubkey = %s, want %s", got, c.AccountIDHex)
		}
	}
}
